services/api/src/utils/middleware/audit.js

```diff
@@ -0,0 +1,79 @@
+const mongoose = require('mongoose');
+
+const { Schema } = mongoose;
+
+const auditSymbol = Symbol('audit');
+
+const schema = new Schema({
+  userId: Schema.Types.ObjectId,
+  collectionName: String,
+  objectId: Schema.Types.ObjectId,
+  object: Schema.Types.Mixed,
+  requestUrl: String,
+  requestMethod: String,
+  requestBody: Schema.Types.Mixed,
+  requestQuery: Schema.Types.Mixed,
+  action: String,
+});
+
+const AuditHistory = mongoose.model('AuditHistory', schema);
+
+function save(ctx, { user, object, collectionName, action }) {
+  return AuditHistory.create({
+    userId: user.id,
+    collectionName: collectionName || object?.collection.collectionName,
+    objectId: object?.id,
+    object,
+    requestUrl: ctx.request.path,
+    requestMethod: ctx.request.method,
+    requestQuery: ctx.request.query,
+    requestBody: ctx.request.body,
+    action: action || ctx._matchedRouteName,
+  });
+}
+
+function setupMiddleware({ getUser = (ctx) => ctx.state.authUser } = {}) {
+  return (ctx, next) => {
+    ctx[auditSymbol] = {};
+
+    return next().then(() => {
+      const auditContext = ctx[auditSymbol] || {};
+      const user = auditContext.user || getUser(ctx);
+
+      const { object, collectionName } = auditContext;
+
+      if (!user) {
+        ctx.logger.warn('[Audit] no user set');
+        return;
+      }
+
+      // if no action(_matchedRouteName) assume the endpoint should not be tracked
+      if (!ctx._matchedRouteName) {
+        return;
+      }
+
+      return save(ctx, {
+        user,
+        object,
+        collectionName,
+      });
+    });
+  };
+}
+
+function setUser(ctx, user) {
+  ctx[auditSymbol].user = user.toObject();
+}
+
+function setObject(ctx, object) {
+  ctx[auditSymbol].collectionName = object.collection.collectionName;
+  ctx[auditSymbol].object = object.toObject();
+}
+
+module.exports = {
+  setup: setupMiddleware,
+  setUser,
+  setObject,
+  save,
```

Comment:
ok it's starting to make more sense now... I don't see `setUser` being used anywhere though so I kind of didn't get how it fit together but I get it now you kind of need to "build up" the audit from different pieces... it seems though there's a lot going on to try to work outside the normal scope of middleware with the extra exports etc... what if it was something more like this:

```js
class Audit {
  constructor(ctx, options = {}) {
    // extract stuff from the context or whatever here
    this.options = options;
  }
  setObject(object) {
    this.object = object;
  }
  setUser(user) {
    this.user = user;
  }
  save() {
    AuditHistory.create({
      ...this.options,
      ...this.object,
      user: this.user.id,
      // ^ this isn't right but you get the idea
    });
  }
} 
module.exports = function auditMiddleware(options) {
  return async (ctx, next) => {
    const audit = new Audit(ctx, options);
    this.state.audit = audit;
    await next();
    if (this.state.authUser) {
      // could do this before `next`, after, or both to allow it to be used with `fetchUser` in any order
      audit.setUser(this.state.authUser);
    }
    audit.save();
  }
}
```

So within a Koa context it can just set defaults and build up the audit pieces along the way then save when it's all done. Outside a Koa context (in a job or something) I think I would just want to use `AuditHistory` directly like any other model...maybe that's what you had in mind already though....

actually thinking about it even the `Audit` step isn't really needed ... you could just do `ctx.state.audit = new AuditHistory()`  and have methods on the model itself like `setUser`, `setObject`, `setContext` that know what information to extract for a given input, then just a normal `save` call when you're all done.

Reply:
you could even say "if `setObject` hasn't been called then grab it from `ctx.body.data`" but that's probably TOO clever :)

Reply:
one open question... this is about DB writes yes? the middleware will cascade so if a GET or POST /search is down the line in the router I'm assuming we don't want an audit trail of reads or searches (or maybe we do??)

Reply:
```
you could even say "if setObject hasn't been called then grab it from ctx.body.data" but that's probably TOO clever :)
```
Its error prune, and some endpoints return 204. But yeah its an option. I rather be verbose (or capture the it from mongoose)

```
one open question... this is about DB writes yes? the middleware will cascade so if a GET or POST /search is down the line in the router I'm assuming we don't want an audit trail of reads or searches (or maybe we do??)
```
Thats up to the implementor, you can easy disable tracking by not naming the route.